### PR TITLE
Add simconfig

### PR DESCRIPTION
**Pull request by julien-bremont**
Adds a `SimConfig` class to pulser to be used by a `Simulation` object. Contains noise settings, number of runs, system temperature... Part of the noisy simulations PR.

### CODE REVIEW

File: pulser/simulation/simconfig.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+from typing_extensions import Literal, get_args
+
+import numpy as np
+import qutip
+
+
+NOISE_TYPES = Literal['doppler', 'amplitude', 'SPAM', 'dephasing']
+MASS = 1.45e-25
+KB = 1.38e-23
+KEFF = 8.7
```
Comment (HGSilveri) [inline pulser/simulation/simconfig.py:28]: Maybe add the units here as comments

File: pulser/simulation/simconfig.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+                    --  eta: Probability of each atom to be badly prepared
+                    --  epsilon: Probability of false positives
+                    --  epsilon_prime: Probability of false negatives.
+            runs (int): Number of runs needed : each run draws a new random
+                noise.
+            samples_per_run (int): Number of samples per noisy run.
+                Useful for cutting down on computing time, but unrealistic.
+            temperature (float): Temperature, set in µK, of the Rydberg array.
+                Also sets the standard deviation of the speed of the atoms.
+            laser_waist (float): Waist of the gaussian laser in global pulses.
```
Comment (HGSilveri) [inline pulser/simulation/simconfig.py:54]: What are the units?

File: pulser/simulation/simconfig.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        # only one noise was given as argument : convert it to a tuple
+        if isinstance(self.noise, str):
+            self.__dict__["noise"] = (self.noise, )
+        for noise_type in self.noise:
+            if noise_type not in get_args(NOISE_TYPES):
+                raise ValueError(cast(str, noise_type) +
+                                 " is not a valid noise type."
+                                 "Valid noise types : " +
+                                 str(get_args(NOISE_TYPES))
+                                 )
```
Comment (HGSilveri) [inline pulser/simulation/simconfig.py:115]: Use f-strings instead of casting to str and summing.

File: pulser/tests/test_simconfig.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+# limitations under the License.
+
+import pytest
+
+from pulser.simulation import SimConfig
+
+
+def test_init():
+    config = SimConfig()
+    str(config)
```
Comment (HGSilveri) [inline pulser/tests/test_simconfig.py:22]: You should initialize a `SimConfig` with custom values and assert that `str(config)` matches what you would expect

**Review by HGSilveri — COMMENTED** [review]
Alright, it looks like it's almost there!

File: pulser/simulation/simconfig.py [hunk 5]
```diff
@@ -62,8 +68,9 @@ class SimConfig:
     epsilon: float = 0.01
     epsilon_prime: float = 0.05
     solver_options: qutip.Options = qutip.Options(max_step=5)
+    spam_dict: dict[str, float] = field(default_factory=dict)
```
Comment (HGSilveri) [inline pulser/simulation/simconfig.py:71]: Why are you adding this one here? Is it the typing?

File: pulser/simulation/simconfig.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from typing import Union
+from typing_extensions import Literal, get_args
```
Comment (HGSilveri) [inline pulser/simulation/simconfig.py:19]: Can you please do the conditional import that is in `sequence.py`? I would rather have users install `typing_extensions` only if they're on Python 3.7, it doesn't make sense to demand that from a user working on Python 3.8.

**Review by HGSilveri — COMMENTED** [review]
Alright, here are a couple more

File: pulser/simulation/simconfig.py (hunk 5 above)
Comment (julien-bremont) [reply pulser/simulation/simconfig.py:71]: Yeah, `mypy` complains that I use `self.spam_dict` otherwise, since I add it in `__post_init__`.

Comment (HGSilveri) [reply pulser/simulation/simconfig.py:80]: Ok, let's hide it from init by setting the `field(init=False, ...)` option. And you can do the same with `doppler_sigma` while you're at it (make it's default be the value corresponding to the default temperature)

**Review by HGSilveri — APPROVED** [review]
Awesome, great job! Congrats on your first PR!

**Pull request merged**